Make default for fields @PolyDet

Currently, the default for fields is `@Det`. I believe `@PolyDet` would be best. Firstly, the default for method and constructor parameters is `@PolyDet`, meaning the following code won't type check by default:

```java
import org.checkerframework.checker.determinism.qual.*;

public class TestDefaults {
    int field;

    public TestDefaults(@PolyDet int field) {
        this.field = field;
    }
}
```
```
TestDefaults.java:8: error: [invalid.field.assignment] cannot assign into @Det field via @PolyDet expression
        this.field = field;
            ^
```

Since the default for constructor parameters is `@PolyDet`, this kind of error would come up in almost any constructor by default.

Secondly, intuitively a field should depend on the determinism of the enclosing class. So, if the enclosing class is not declared `@Det`, the most logical choice would be for the fields to be `@PolyDet` themselves.

--- a/checker/src/main/java/org/checkerframework/checker/determinism/DeterminismAnnotatedTypeFactory.java
+++ b/checker/src/main/java/org/checkerframework/checker/determinism/DeterminismAnnotatedTypeFactory.java
@@ -957,4 +957,26 @@ public class DeterminismAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
             return super.isSubtype(subAnno, superAnno);
         }
     }
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Defaults all fields in a class with an implicit qualifier parameter to {@code PolyDet}.
+     *
+     * @param elt Element whose type is {@code type}
+     * @param type where the defaults are applied
+     */
+    @Override
+    protected void applyQualifierParameterDefaults(Element elt, AnnotatedTypeMirror type) {
+        if (elt == null || elt.getKind() != ElementKind.FIELD || type.isAnnotatedInHierarchy(DET)) {
+            super.applyQualifierParameterDefaults(elt, type);
+            return;
+        }
+
+        TypeElement enclosingClass = ElementUtils.enclosingClass(elt);
+        Set<AnnotationMirror> tops = getQualifierParameterHierarchies(enclosingClass);
+        if (AnnotationUtils.containsSameByClass(tops, NonDet.class)) {
+            type.addAnnotation(POLYDET);
+        }
+        super.applyQualifierParameterDefaults(elt, type);
+    }
 }

--- a/checker/src/main/java/org/checkerframework/checker/determinism/DeterminismAnnotatedTypeFactory.java
+++ b/checker/src/main/java/org/checkerframework/checker/determinism/DeterminismAnnotatedTypeFactory.java
@@ -967,7 +967,10 @@ public class DeterminismAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
      */
     @Override
     protected void applyQualifierParameterDefaults(Element elt, AnnotatedTypeMirror type) {
-        if (elt == null || elt.getKind() != ElementKind.FIELD || type.isAnnotatedInHierarchy(DET)) {
+        if (elt == null
+                || elt.getKind() != ElementKind.FIELD
+                || ElementUtils.isStatic(elt)
+                || type.isAnnotatedInHierarchy(DET)) {
             super.applyQualifierParameterDefaults(elt, type);
             return;
         }

--- a/checker/tests/determinism/DefaultField.java
+++ b/checker/tests/determinism/DefaultField.java
@@ -1,0 +1,23 @@
+import org.checkerframework.checker.determinism.qual.Det;
+import org.checkerframework.checker.determinism.qual.NonDet;
+import org.checkerframework.checker.determinism.qual.PolyDet;
+import org.checkerframework.framework.qual.HasQualifierParameter;
+
+@HasQualifierParameter(NonDet.class)
+public class DefaultField {
+    Object field;
+
+    @PolyDet Object getField(@PolyDet DefaultField this) {
+        return field;
+    }
+
+    @PolyDet Object getField2() {
+        return field;
+    }
+
+    void test() {
+        // :: error: (assignment.type.incompatible)
+        @Det Object o = field;
+        @NonDet Object o2 = field;
+    }
+}

--- a/checker/tests/determinism/DefaultField.java
+++ b/checker/tests/determinism/DefaultField.java
@@ -5,6 +5,7 @@ import org.checkerframework.framework.qual.HasQualifierParameter;
 
 @HasQualifierParameter(NonDet.class)
 public class DefaultField {
+    static Object staticField;
     Object field;
 
     @PolyDet Object getField(@PolyDet DefaultField this) {
@@ -19,5 +20,6 @@ public class DefaultField {
         // :: error: (assignment.type.incompatible)
         @Det Object o = field;
         @NonDet Object o2 = field;
+        @Det Object o3 = staticField;
     }
 }

--- a/checker/tests/determinism/DefaultField2.java
+++ b/checker/tests/determinism/DefaultField2.java
@@ -1,0 +1,22 @@
+import org.checkerframework.checker.determinism.qual.Det;
+import org.checkerframework.checker.determinism.qual.NonDet;
+import org.checkerframework.checker.determinism.qual.PolyDet;
+
+public class DefaultField2 {
+    static Object staticField;
+    Object field;
+
+    @Det Object getField(@Det DefaultField2 this) {
+        return field;
+    }
+
+    @PolyDet Object getField2(@PolyDet DefaultField2 this) {
+        return field;
+    }
+
+    void test(@Det DefaultField2 this) {
+        @Det Object o = field;
+        @NonDet Object o2 = field;
+        @Det Object o3 = staticField;
+    }
+}

